### DISCUSSION TITLE
Populating feeds by last published entries, when last published ≠ `date`?

### DISCUSSION THREAD

**Discussion by j9t** [11ty/eleventy#2716] [Q&A] [answered]
In an Eleventy project, I use the front matter `date` field to mark the publication date of entries—with most entries being published _on other sites_.

For the respective site feed, I’d like sort the last _n_ entries being published _on the project_.

That means, I cannot populate and sort the feed by `date`, because the publication date on the project equals or is later than the front matter `date` which is the original publication date.

---

At the moment, I help myself by sorting the feed reverse-chronologically by front matter date, accepting that only those entries that happen to fall into the date range will be featured, somewhere in the feed. (As the site contains thousands of entries—and growing—, I could not just add all entries to the feed.)

Ideally, there would be some reliable and easy-to-maintain way to use the `date` field to indicate the post publication date, and something else (like a “last edited” date) to use for inclusion in the feed.

I’m aware that `date` has values like “Last Modified” and such—but apart from those not seeming to work reliably for me (TBC, and to be reported as an issue if confirmed), there’s value to reserve this field for the purpose I’m using it for now (publication date of the resources the entries point to). I also _believe_ there were plans (or requests) to have another date field, for perhaps this use case.

But what do I miss? Is there some elegant solution to this? Or what else do you notice? (Thank you!)

**Comment by btrem** [comment]
Perhaps I don't understand the issue, but you can add a field to your front matter:
```
---
date: 2023-01-02
date-edited: 2023-01-16
title: My Post
---
Lorem ipsum....
```

**Reply by j9t** [reply]
That would require to maintain two dates, which I’d like to avoid. (I also believe that `date-edited` would not be treated like `date` would, possibly requiring some different handling.)

Maybe a simpler way of putting the question (problem) is, is there a way to sort by when a file was published (or modified)—without manual work, and without resorting to `date`?

**Reply by btrem** [reply]
I don't really understand the question, then. You want to sort by a date but you don't want to include one of them? If there are two relevant dates -- published and edited -- and you want to sort by both, then you'll have to include both in some manner.

**Reply by j9t** [reply]
> If there are two relevant dates -- published and edited -- and you want to sort by both, then you'll have to include both in some manner.

No, neither practically (`date` allows [values like “Last Modified”](https://www.11ty.dev/docs/dates/)—but which, if I’m not mistaken, doesn’t yet fully work with my particular use case) nor theoretically (something like file-editing information could be taken “automatically,” and if the sky was the limit, a system could also parse referenced sources and detect when these were published).

That fuels my hope that it may indeed be possible to sort by two time-related values without maintaining two.

**Reply by btrem** [reply]
> > If there are two relevant dates -- published and edited -- and you want to sort by both, then you'll have to include both in some manner.
> 
> No, neither practically (`date` allows [values like “Last Modified”](https://www.11ty.dev/docs/dates/)—but which, if I’m not mistaken, doesn’t yet fully work with my particular use case)

But if `date` did allow a value of "Last Modified," then it would no longer have the published date, so I don't see how that would solve your problem. You need two dates, but you don't want to put two dates in your front matter. There might be other solutions, but they will likely involve some coding work, or might be brittle, or both.

> nor theoretically (something like file-editing information could be taken “automatically,” and if the sky was the limit, a system could also parse referenced sources and detect when these were published).

You could look at microformats `rel=syndicated` and `u-syndicated`. But that would still require you to link the copy on your site with the one on another site, either in the front matter or in the content. And it would mean you'd be relying on the other site -- the syndicating site, to use microformats terminology -- to reliably provide the data you need, i.e., the date that the site published  your article.

**Reply by j9t** [reply]
Yes!—I don’t have much hope anymore for a simple and elegant solution :)

**Comment by pdehaan** [comment, marked as answer]
I think I'm still confused.
It sounds like you want to have two dates per post:
- "published date" &mdash; with most entries being published on other sites.
- "edited date" &mdash; when the content was added to your site.

How often does a "published date" change?
Why not hard code the external "published date" in your front matter (like `publishedDate: 2022/04/14`) and then use something like `date: git Last Modified` (or whatever it is on https://www.11ty.dev/docs/dates/) for when the content was posted on your site?

**Reply by pdehaan** [reply]
Clumsy example at https://github.com/pdehaan/11ty-2716

Templates roughly look like this (where `{% time %}` is a silly custom shortcode which wraps it in a `<time>` tag):

```liquid
---
title: FoUr
date: Last Modified
publishedDate: 2022-04-24
---

# {{ title }}

{% time page.date %}

{% time publishedDate %}
```

### src/page_date.liquid

Sort the `collections.pages` collection by `data.page.date` (which would be your local last modified date):

```liquid
---
title: PaGeS by page.date
---

<h1>{{ title }}</h1>

{%- assign pages_by_page_date = collections.pages | sort: "data.page.date" -%}

{% for p in pages_by_page_date reversed %}
<article>
  <h2>{{ p.data.title }}</h2>
  <p>page date: {% time p.data.page.date %}</p>
  <!-- <p>publish date: {% time p.data.publishedDate %}</p> -->
</article>
{% endfor %}
```

### src/publish_date.liquid

Sort the `collections.pages` collection by hard-coded `data.publishedDate` front matter (which is when the content was something something "external site"):

```liquid
---
title: PaGeS by publishedDate
---

<h1>{{ title }}</h1>

{%- assign pages_by_published_date = collections.pages | sort: "data.publishedDate" -%}

{% for p in pages_by_published_date reversed %}
<article>
  <h2>{{ p.data.title }}</h2>
  <!-- <p>page date: {% time p.data.page.date %}</p> -->
  <p>publish date: {% time p.data.publishedDate %}</p>
</article>
{% endfor %}
```

So a bit LiquidJS specific since Nunjucks doesn't have a built-in `sort` filter, last I checked. But easy enough to refactor into the .eleventy.js config file as custom pre-sorted collections sorted by whatever dimension(s) you want.

**Reply by j9t** [reply]
This is it, I was wondering about (and shying away from) using `Last Modified` with `date`, and setting the publications date on a different field.

However, switching to this approach (which so far looks promising) I run into a major issue which I can’t seem to overcome (even after consulting other of your awesome advice, as with [#869](https://github.com/11ty/eleventy/issues/869)):

While Eleventy parses `date: Last Modified` fine, I get “NaN” for anything like “data-published: 2023-01-28”—presumably manipulating it through `htmlDateString` or other filters doesn’t work on the correct object?

Working through said threads I can’t figure out how to convert “date-published” to a “proper” date to filter and work with just as with “date”. What do I miss?

**Reply by pdehaan** [reply]
NaN looks suspicious.
Try `datePublished` or `date_published` instead in case it’s trying to do some maths with the hyphen in the front matter variable name.

**Reply by j9t** [reply]
Eh. Yes. This essentially works now—but with one other issue: On my (paginated Nunjucks) index page I can’t seem to reliably sort by this:

```
{% for entry in entries | sort(attribute='data.date_publication') | reverse %}
```

What’s interesting is that this _does_ seem to work when there’s no pagination, like on tag pages. After the switch to go by `date_publication` it seems sorting only works per page, not by the entire collection of posts.

Does this ring familiar as well? As previously, `{% for entry in entries %}` worked without any filters, I wonder if Eleventy does something special with the `date` field, which is now unavailable?

(Thank you so much for the help…!)

**Reply by pdehaan** [reply]
Yeah, I think 11ty sorts collections by `date` by default.
I'd probably just create custom collections in your .eleventy.js config file (something like `entries_date_publication`) and pre-sort them however you want, which should work for pagination or loops or whatever else. 🤷

**Reply by pdehaan** [reply]
Oh, maybe I understand what you're saying...

If `entries` above is your paginated `alias` then yeah, I can understand why that'd not work like one might expect (due to order of pagination and sorting).

If I hypothetically have `129` blog entries, and I have a pagination size of `19`, the first page will give me 19 of the items from my collection/data in whatever default sort order it has (probably [default] sorted by `date`, or whatever -- unless it's a custom collection).
Now, when you use a `{% for entry in entries %}` loop, you're just looping over those 19 entries in the first pagination "chunk" and then doing secondary sorting on the subset of the pagination data, not the whole set. So essentially you're sorting 19 random entries by `date_publication` front matter key).

If any of that nonsense is roughly accurate, you'd have two options.
1. pre-sort the collection/data so pagination "just works"(tm) using a custom collection/global data file.
    - PROS: Relatively easy.
    - CONS: An extra collection which might only be needed in 1 place.
3. use something like the pagination [`before` callback](https://www.11ty.dev/docs/pagination/#the-before-callback) to pre-sort the data _before_ calculating the pagination `size` chunking.
    - PROS: Fewer custom collections.
    - CONS: Would require JavaScript front matter (or possibly .JS template data file to do the `before` sorting/filtering).

**Reply by j9t** [reply]
Thanks @pdehaan! Testing option 1)…

**Reply by j9t** [reply]
I think I got this to work—thank you!

It doesn’t feel great yet—I observe inconsistencies how Eleventy handles `date: "Last Modified"` in front matter vs. in folder JSON files, and something like `eleventyConfig.addCollection("postsByPublicationDate", function(collectionApi) {}` seems to create a phantom “postsByPublicationDate” tag (which I could exclude manually)—, but this should go elsewhere, and overall, it’s much better.

Thanks @pdehaan!
